src/switchImplIntf.ml

```diff
@@ -0,0 +1,98 @@
+open Import
+
+module Lsp : sig
+  val switch : LanguageClient.t -> TextDocument.t -> string array Promise.t
+end = struct
+  let switch client document =
+    let targetFileName : string array Promise.t =
+      LanguageClient.sendRequest client ~meth:"ocamllsp/didSwitchImplIntf"
+        ~data:(document.TextDocument.fileName : string)
+        ()
+    in
+    targetFileName
+end
+
+module Fallback : sig
+  val switch : string -> string array Promise.t
+end = struct
+  let impl = [ ".ml"; ".mll"; ".mly"; ".re" ]
+
+  let intf = [ ".mli"; ".rei" ]
+
+  let switchTo exts withoutExt =
+    Promise.List.find_map exts ~f:(fun ext ->
+        let targetFileName = withoutExt ^ ext in
+        let open Promise.O in
+        Fs.exists targetFileName >>| function
+        | true -> Some targetFileName
+        | false -> None)
+
+  (* let switch fileName =
+     let ext = Filename.extension fileName in
+     let withoutExt = Filename.remove_extension fileName in
+     if List.mem ext intf then
+       switchTo impl withoutExt
+     else if List.mem ext impl then
+       switchTo intf withoutExt
+     else
+       Promise.return None *)
+
+  let switch _ = failwith "not implemented" (* FIXME *)
+end
+
+let requestSwitch ~client ~capabilities document =
+  let showFile targetFileName =
+    Window.showTextDocument (`Uri (Uri.file targetFileName)) |> ignore
+  in
+
+  let onPromiseError error =
+    let errorMessage = Node.JsError.ofPromiseError error in
+    let showSwitchError error =
+      message `Error "Error when switching implementation/interface: %s" error
+    in
+    showSwitchError errorMessage;
+    Promise.return ()
+  in
+
+  let filesToSwitchTo =
+    match (client, capabilities) with
+    | Some client, Some capabilities
+      when OcamlLsp.handleSwitchImplIntf capabilities ->
+      Lsp.switch client document
+    | _ ->
+      log "using fallback mechanism to switch implementation/interface";
+      Fallback.switch document.TextDocument.fileName
+  in
+
+  let open Promise.O in
+  filesToSwitchTo >>= fun arr -> match Array.to_list arr with
+  | [] -> assert false
```

Comment:
Is it not possible for the language server to return an empty array?

Reply:
At the moment of writing this code, it wasn't possible, but since we're [discussing](https://github.com/ocaml/ocaml-lsp/pull/233#discussion_r472937245) some behavior change on ocaml-lsp side, this is subject to change.